### ISSUE TITLE
Introduce CI

### ISSUE THREAD

**Issue by Bios-Marcel** [Bios-Marcel/top-down-pew-pew#6]
To assure that mistakes don't end up in the master branch, we need to have a second branch called `develop` or something along the lines of that.

Everything pushed to `develop` will trigger a build and the contributors will be alarmed on failure. If `develop` is building successfully, it can be merged into `master`.

The CI setup can happend without `develop` existing.

For Windows https://appveyor.com should be used.

For Linux and Mac OS https://travis-ci.org and https://circleci.com can be used.

Todo:
- [ ] Setup a `develop` branch
  - [ ] Configure repository to only allow merging if CI doesn't fail
- [ ] Builds for Mac OS
- [ ] Builds for Linux x64
- [ ] Builds for Windows x64

**Comment by kaskelotti** [comment]
Hey,

I'm interested in helping you with this one. I have questions / ideas related to your CI setup.

1. To me it sounds like your plan is that all commits are made to one branch and then merged to master, if they pass the build. Why not instead use separate branch for each feature / bug and us pull requests (PR) for merging to master? Then all commits to your project would need to come through PRs.

2. The purpose CI build is that it happens "behind the scenes", i.e. the CI server will pull the sources from GitHub and run the build script. There is no need to specify separate environments for Windows/Mac/Linux. Unless your project has some environment dependencies in tests. If there are no environment related dependencies, I'd recommend going with Travis CI only, as it's extremely simple to integrate with GitHub.

**Comment by Bios-Marcel** [comment]
The idea is to produce artifacts for every platform aswel.
I am not sure if Godot can always produce working artifacts by crosscompiling?

However, creating branches for every features is too annoying to us. Maybe later on in the future we might change our workflow if we notice that we have too many problems.

Currently this is the plan for our workflow:

Make commits to `develop` and merge into `master` if it's fine. 
Make PRs onto `develop` and merge into `master` if all is fine.

Merges might not happen instantly, we haven't really thought of a release model yet, as we aren't even close to being feature complete, since we just started.

**Comment by Bios-Marcel** [comment]
I think you could also have `develop`, `master` and feature branches though. I am not sure if I'd want to create a branch for every bugfix.

**Comment by Bios-Marcel** [comment]
There is now a develop branch.

**Comment by kaskelotti** [comment]
I did some research on this.

None of the said CI services support Godot/ GDSCript out of the box. All of them seem to have some sort of before block where one can define what software needs to be installed before the build can begin. For example Travis has `before_install` block where one can define packages to be installed from Apt(Linux) or brew(OSX).

OSX with brew allows installing Godot from package manager. For Windows Godot has a scoop package available, but Appveyor has only support for chocolatey package manager, where Godot is not found. Godot web page does not describe any options for Linux package managers, and I could not find it in Apt.

There's a chance that some custom Apt repository has Linux version available, though. In addition at least Travis has an option to get a package with wget, but documentation only discusses using .deb packages. The Godot engine would need to be unzipped and added to PATH to be usable. Otherwise it would require some other mean for installing the Godot engine, perhaps a pre-defined image could be used, and TBH I have no ideas on how to get this sort of setup ready.

And, only Linux version seems to have headless version available. Unless the others have a CLI available this will get tricky.

Godot documentation describes cross compilation to be possible only from Linux to MacOS target, but this is said to be unstable. Thus, it would probably require separate environment for each targets.

Long story short: With Travis this might be possible to setup for OSX with the following steps
- Define OSX environment use
- Get the Godot engine in `before_install` block
- Use CLI (if available for OSX) to run the build/tests
- Store the build artifacts

Considering these obstacles, the task of introducing CI is not going to be as easy as I originally though. Unfortunately I cannot commit enough time and effort to resolve this. I hope some other hackatonist will be able to participate and get this done.

P.S. There nothing wrong with your branching policy of having only master and development, if you think it will suit you. And it's always easy to scale up and start using PRs if they become necessary.

**Comment by Bios-Marcel** [comment]
I see. Thanks for your research effort. Also, we have tried the "PR for everything" approach. It's fine, that way nothing gets uploaded without being seen.
